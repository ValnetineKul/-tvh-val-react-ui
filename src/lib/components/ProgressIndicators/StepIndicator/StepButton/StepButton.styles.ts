import { makeStyles } from '../../../../../themes/core';

const useStyles = makeStyles<void, 'icon' | 'disapledIcon' | 'label' | 'alternativeLabel' | 'hiddenLabel'>({
  name: 'StepButton',
})((theme, _, classes) => ({
  icon: {
    display: 'flex',
    justifyContent: 'center',
    width: `calc( ${theme.layoutSpacing['spacing/400']} + 3px)`,
    height: theme.layoutSpacing['spacing/400'],
    transform: 'skewX(-10deg)',
  },
  disapledIcon: {
    backgroundColor: theme.color['bg/disabled'],
  },
  stepButton: {
    padding: theme.layoutSpacing['spacing/000'],
    margin: theme.layoutSpacing['spacing/000'],
    '&:hover': {
      [`& .${classes.icon}`]: {
        backgroundColor: theme.color['bg/surfacePrimary/hover'],
      },
      [`& .${classes.label}`]: {
        textDecoration: 'underline',
      },
    },
    '&:focus': {
      [`& .${classes.icon}`]: {
        boxShadow: `0px 0px 0px 4px ${theme.color['border/onSurfaceCurrent/focus']}`,
      },
    },
  },
  label: {
    minWidth: 0,
  },
  iconContainer: {
    position: 'relative',
    zIndex: 10,
  },
  labelContainer: {
    minWidth: 0,
    [`& .${classes.alternativeLabel}`]: {
      marginTop: theme.layoutSpacing['spacing/200'],
    },
  },
  rootLabel: {
    minWidth: 0,
  },
  truncatedLabel: {
    display: 'block',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    [`&.${classes.hiddenLabel}`]: {
      display: 'flex',
      paddingBottom: theme.layoutSpacing['spacing/200'],
      marginBottom: `-${theme.layoutSpacing['spacing/200']}`,
    },
  },
  tooltip: {
    paddingTop: theme.layoutSpacing['spacing/200'],
    marginTop: `-${theme.layoutSpacing['spacing/200']}`,
  },
  alternativeLabel: {},
  hiddenLabel: {},
}));

export default useStyles;
