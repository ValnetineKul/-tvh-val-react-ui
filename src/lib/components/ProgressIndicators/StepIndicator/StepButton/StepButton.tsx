import type { FC, MouseEventHandler } from 'react';
import React, { useEffect, useRef, useState } from 'react';
import StepLabel from '@mui/material/StepLabel';
import { default as MuiStepButton } from '@mui/material/StepButton';
import type { StepIconProps } from '@mui/material/StepIcon';
import Surface from '../../../Surfaces/Surface';
import { Check } from '../../../Icon/icons/functional';
import Icon from '../../../Icon';
import Typography from '../../../Typography';
import Tooltip from '../../../Tooltip';
import type { ButtonBaseProps } from '../../../ButtonBase';
import { useButtonBase } from '../../../ButtonBase';
import useStyles from './StepButton.styles';

export interface StepButtonProps extends ButtonBaseProps {
  label: string;
  hiddenLabel?: boolean;
  disabled?: boolean;
  onClick?: MouseEventHandler<HTMLButtonElement | HTMLAnchorElement>;
  className?: string;
}

const StepButton: FC<StepButtonProps> = ({ label, hiddenLabel = false, onClick, className, ...props }) => {
  const { classes, cx } = useStyles();

  const labelRef = useRef<HTMLSpanElement>(null);
  const [shouldShowTooltip, setShouldShowTooltip] = useState(false);

  useEffect(() => {
    const compareSize = () => {
      setShouldShowTooltip(
        !!(label && labelRef && labelRef.current && labelRef.current.scrollWidth > labelRef.current.clientWidth)
      );
    };

    compareSize();

    window.addEventListener('resize', compareSize);
    return () => {
      window.removeEventListener('resize', compareSize);
    };
  }, [label]);

  const buttonBaseProps = useButtonBase(props);

  const StepIconComponent = ({ active, completed, icon }: StepIconProps) => {
    const renderInnerComponent = () => {
      return completed ? (
        <Icon icon={Check} size="sm" />
      ) : (
        <Typography variant="body300" weight="emphasis" component="span">
          {icon}
        </Typography>
      );
    };

    if (active || completed) {
      return (
        <Surface color="Primary" className={classes.icon}>
          {renderInnerComponent()}
        </Surface>
      );
    }

    return <span className={cx(classes.icon, classes.disapledIcon)}>{renderInnerComponent()}</span>;
  };

  return (
    <Tooltip
      label={label}
      disableHoverListener={!shouldShowTooltip}
      className={cx(classes.truncatedLabel, classes.tooltip, { [classes.hiddenLabel]: hiddenLabel })}
    >
      <MuiStepButton onClick={onClick} className={cx(classes.stepButton, className)} {...buttonBaseProps}>
        <StepLabel
          StepIconComponent={StepIconComponent}
          classes={{
            root: classes.rootLabel,
            label: classes.label,
            iconContainer: classes.iconContainer,
            labelContainer: classes.labelContainer,
            alternativeLabel: classes.alternativeLabel,
          }}
        >
          {!hiddenLabel && (
            <Typography variant="body400" component="span" ref={labelRef} className={classes.truncatedLabel}>
              {label}
            </Typography>
          )}
        </StepLabel>
      </MuiStepButton>
    </Tooltip>
  );
};
export default StepButton;
