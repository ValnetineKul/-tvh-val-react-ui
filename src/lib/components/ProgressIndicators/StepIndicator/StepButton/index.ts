import StepButton from './StepButton';

export type { StepButtonProps } from './StepButton';
export default StepButton;
