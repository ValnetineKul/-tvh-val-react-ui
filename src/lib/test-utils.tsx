import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
