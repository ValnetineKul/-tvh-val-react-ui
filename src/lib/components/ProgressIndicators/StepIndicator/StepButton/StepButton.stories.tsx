import React from 'react';
import { createMeta } from '../../../../story-utils';
import type { StoryTemplate } from '../../../../story-utils';
import StepIndicator from '../StepIndicator';
import Step from '../Step';
import StepButton from './StepButton';
import { buttonBaseStoryArgs } from '../../../ButtonBase';

type Props = React.ComponentProps<typeof StepButton>;
type Args = Omit<Props, 'label' | 'to' | 'disabled'>;

const Template = ((args: Args) => {
  return (
    <StepIndicator activeStep={1}>
      <Step completed>
        <StepButton {...args} label="Completed step (link)" href="/whereToGoTo" />
      </Step>
      <Step completed={false}>
        <StepButton {...args} label="Active step" />
      </Step>
      <Step>
        <StepButton {...args} label="Future step" disabled />
      </Step>
    </StepIndicator>
  );
}) as StoryTemplate<Args>;

export const Primitive = Template.bind({});
Primitive.args = {
  hiddenLabel: false,
};
Primitive.storyName = 'StepButton';

export default createMeta({
  component: StepButton,
  title: 'Components/ProgressIndicators/StepIndicator/StepButton',
  argTypes: {
    label: {
      control: {
        disable: true,
      },
    },
    disabled: {
      control: {
        disable: true,
      },
    },
    ...buttonBaseStoryArgs,
  },
});
