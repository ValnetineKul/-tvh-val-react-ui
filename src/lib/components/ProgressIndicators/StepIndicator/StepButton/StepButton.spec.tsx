import React from 'react';
import { Link } from 'react-router-dom';
import { render, userEvent, screen } from '../../../../test-utils';
import StepButton from './StepButton';

describe('StepButton', () => {
  it('Should render correctly', () => {
    const { container } = render(<StepButton label="Label" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should disable button if disabled prop is true', () => {
    render(<StepButton label="Label" disabled />);
    const button = screen.getByRole('button', { name: 'Label' });
    expect(button).toBeDisabled();
    expect(button.className).toMatch('Mui-disabled');
  });

  it('Should callback on button click', () => {
    const onClick = jest.fn();
    render(<StepButton label="Label" onClick={onClick} />);
    const button = screen.getByRole('button', { name: 'Label' });
    userEvent.click(button);
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it('Should render HTML link tag "a" if "href" prop is filled', () => {
    render(<StepButton label="Label" href="/href" />);
    expect(screen.getByRole('link', { name: 'Label' })).toBeInTheDocument();
  });

  it('Should render Link from "react-router-dom" library if "to" prop is filled', () => {
    render(<StepButton label="Label" component={Link} to="/to" />);
    expect(screen.getByRole('link', { name: 'Label' })).toBeInTheDocument();
  });

  it('Should hide label if hiddenLabel prop is true', () => {
    render(<StepButton hiddenLabel label="Label" />);
    expect(screen.getByRole('button')).toBeInTheDocument();
    expect(screen.queryByText('Label')).not.toBeInTheDocument();
  });
});
