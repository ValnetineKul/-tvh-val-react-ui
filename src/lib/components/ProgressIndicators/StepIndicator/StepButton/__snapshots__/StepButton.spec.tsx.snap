// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepButton Should render correctly 1`] = `
<div
  aria-label="Label"
  class="tss-2-Typography-weightRegular-ref tss-tls3jp-Typography-root-Typography-weightRegular-Tooltip-wrapper-StepButton-truncatedLabel-StepButton-tooltip-Typography-body400"
  data-mui-internal-clone-element="true"
>
  <button
    class="MuiButtonBase-root MuiStepButton-root tss-12w7rqv-StepButton-stepButton mui-mjpk2e-MuiButtonBase-root-MuiStepButton-root"
    tabindex="0"
    type="button"
  >
    <span
      class="tss-kicety-StepButton-rootLabel MuiStepLabel-root mui-ascpo7-MuiStepLabel-root"
    >
      <span
        class="tss-19uxhif-StepButton-iconContainer MuiStepLabel-iconContainer mui-vnkopk-MuiStepLabel-iconContainer"
      >
        <span
          class="tss-4-StepButton-icon-ref tss-1cshna-StepButton-icon-StepButton-disapledIcon"
        >
          <span
            class="tss-2-Typography-weightEmphasis-ref tss-ka481l-Typography-root-Typography-weightEmphasis-Typography-body300"
          />
        </span>
      </span>
      <span
        class="tss-f51pl8-StepButton-labelContainer MuiStepLabel-labelContainer mui-16ubnlw-MuiStepLabel-labelContainer"
      >
        <span
          class="tss-2u1xn3-StepButton-label tss-4-StepButton-label-ref MuiStepLabel-label mui-rngahu-MuiStepLabel-label"
        >
          <span
            class="tss-2-Typography-weightRegular-ref tss-69dlm5-Typography-root-Typography-weightRegular-StepButton-truncatedLabel-Typography-body400"
          >
            Label
          </span>
        </span>
      </span>
    </span>
  </button>
</div>
`;
